### ISSUE TITLE
About mIOU

### ISSUE THREAD

**Issue by callmelaoyi** [qjadud1994/DRS#7]
Thank your brilliant Idea! I've trained a deeplabv3+ by your official pseudo labels and your network. However, I evaluate the mIOU of val_cls.txt that only achieve about 67.1 without dCRF. All of the hyperparameter is arranged to such as  yours.  I wanna know why my mIOU is lower than yours.
Thank your for your reply. Best wishes with you.

**Comment by qjadud1994** [comment]
Just a few days ago, we evaluated the performance of DeepLabv3.
Even the seed is fixed, so the performance will be reproduced if the implementation environment is the same.

First, we will check if we've properly uploaded the official pseudo labels.

Can you tell me about your implementation environment?

**Comment by callmelaoyi** [comment]
gpu: RTX 3090 number: 3
python: 3.8
cudatoolkit: 11.0
pytorch: 1.7.1
About pseudo labels:  I generated those with official weight(Refinement Network).
About deeplabv3+: I trained network by run.sh that corrected image_root.
The rest of arrangement didn't changed.

**Comment by callmelaoyi** [comment]
I observe some saliency maps and official pseudo labels. For example, I notice the image  named "2007_000032" and find that something should be foreground in the saliency map towards the lower left corner howerver in the pseudo label is regarded as background. So I guess if the thredhold of  background is 50 or not?

**Comment by qjadud1994** [comment]
ok.

First, please check the performance of DeepLab-V3+ trained using [refined_pseudo_segmentation_labels/](https://drive.google.com/drive/folders/1IS9_YCrRJwz3c7y3KwTET2_dYUPlNYo6?usp=sharing) we provided.

and then, we will check the pseudo label generation procedure.

**Comment by callmelaoyi** [comment]
Thank you for your reminder. Last night I trained V3 network with your official pseudo labels. But the mean IOU was still 67.4. So I don't think there are problems in the pseudo labels. 
There are output of console.
---------------
Overall Acc: 0.916536
Mean Acc: 0.806906
FreqW Acc: 0.854189
Mean IoU: 0.674887

Model saved as checkpoints/best_deeplabv3plus_resnet101_voc_os16.pth
----------------

**Comment by qjadud1994** [comment]
Ok, thank you for sharing the result.

I still could not catch where the performace gap came from.

So, lastly, could you check the mIoU result from [our DeepLabV3 weight](https://github.com/qjadud1994/DRS#training-the-deeplab-v3-using-pseudo-labels)?

**Comment by callmelaoyi** [comment]
Well, thank you very much for your patient reply! I checked the mIOU from official DeepLabV3 weight. And the mIOU in validation list(1449) is 67.4. I wander if how to get that result(70.4).

**Comment by qjadud1994** [comment]
Sorry for your confusion.
Finally, I figured out the cause of the issue.

That is `crop_val` argument in DeepLabV3 code.
During our code refactoring, we forgot to enable `crop_val` option.

Note that the [original DeepLabV3 implementation](https://github.com/VainF/DeepLabV3Plus-Pytorch#34-testing) set `crop_val=True`.

We update our source code and please check it again.

![image](https://user-images.githubusercontent.com/20681582/124342095-44f3c500-dbfc-11eb-92d5-cffbb5e041ff.png)

**Comment by callmelaoyi** [comment]
Your work was acclaimed as a masterpiece.
Now to further confuse the issue, is this techniques called "crop_val" that could be compared with the other state-of-the-arts method allowed in the formal paper?

**Comment by qjadud1994** [comment]
In WSSS fields, the main goal is to generate high-quality pseudo segmentation labels.

And, the purpose of using the segmentation network (e.g, DeepLab) is measure the quality of the pseudo labels.
Therefore, most of existing WSSS methods strictly follow the training recipe of the segmentation network.

Since enabling `crop_val` is a default setting of the DeepLabV3, the disabling `crop_val` is just our mistake, not related to the   contibutions.

(For the update DeepLabV3 result, we originally used `crop_val=True` for both training and evaluation. However, during the code refactorying, we just forgot to set `crop_val=True`. )

**Comment by callmelaoyi** [comment]
Uhhhh,I got to know it. :) Thank you for your  patience answers again. I'm interested in WSSS. I sincerely want to cooperate with your team. (This is not modest one but honest one,)

**Comment by qjadud1994** [comment]
Thanks to you, we can fix an issue in the DeepLabV3 code.

Unfortunately, I don't have a plan to research WSSS because I'm busy with my current works in our company :( 

I hope to have the opportunity to collaborate on WSSS research with you in the future.
